File: parth/Week1/LinkedListCycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (head == nullptr || head->next == nullptr)
        {
            return false;
        }
        ListNode* slow = head;
        ListNode* fast = head;
        while(fast != nullptr)
        {
            slow = slow->next;
            fast = fast->next!=nullptr ? fast->next->next : nullptr;
            if (slow == fast)
            {
                return true;
            }
        }
        return false;
    }
};

File: Manali/LinkedList/LinkedListCycle.py
# "Floyd's Cycle-Finding Algorithm" or "Tortoise and Hare Algorithm".
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head: return False
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow==fast: return True
        return False
    
TC: O(n)
SC: O(1)
        